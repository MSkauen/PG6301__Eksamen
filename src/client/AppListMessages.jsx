import React from "react";
import { LoadingView } from "./LoadingView";
import { Link } from "react-router-dom";
import { useLoading } from "./useLoading";
import { ErrorView } from "./ErrorView";

export function AppListMessages({ messageApi }) {
  const { error, loading, data, reload } = useLoading(async () =>
    messageApi.listMessages()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }
  if (loading) {
    return <LoadingView />;
  }

  return (
    <>
      <h1>List messages</h1>
      {data.map(({ id, subject }) => (
        <li key={id}>
          <Link to={`/edit?id=${id}`}>{subject}</Link>
        </li>
      ))}
    </>
  );
}
