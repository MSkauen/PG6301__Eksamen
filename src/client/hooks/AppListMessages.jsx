import React from "react";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";

export function AppListMessages({ messageApi }) {
  const { data: messages, error, loading, reload } = useLoading(
    async () => await messageApi.listMessages()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }
  if (loading || !messages) {
    return <LoadingView />;
  }

  return (
    <>
      <h1>List messages</h1>
      {messages.map(({ id, subject }) => (
        <li key={id}>
          <Link to={`/messages/${id}/edit`}>{subject}</Link>
        </li>
      ))}
    </>
  );
}
