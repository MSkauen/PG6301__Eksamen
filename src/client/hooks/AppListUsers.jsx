import React from "react";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";
import { Link } from "react-router-dom";

export function AppListUsers({ userApi }) {
  const { data: users, error, loading, reload } = useLoading(
    async () => await userApi.listUsers()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }
  if (loading || !users) {
    return <LoadingView />;
  }

  return (
    <>
      <h1>List users</h1>
      {users.map(({ id, firstName }) => (
        <li key={id}>
          <Link to={`/users/${id}/edit`}>{firstName}</Link>
        </li>
      ))}
    </>
  );
}
