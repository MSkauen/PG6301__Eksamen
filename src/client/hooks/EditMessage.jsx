import React, { useState } from "react";
import { useParams } from "react-router";
import { LoadingView } from "../components/LoadingView";
import { InputField } from "../components/InputField";
import { ErrorView } from "../components/ErrorView";
import { useLoading } from "../lib/useLoading";

function EditMessageForm({ message }) {
  const [subject, setSubject] = useState(message.subject);
    const [recipient, setRecipient] = useState(message.recipient);
    const [content, setContent] = useState(message.content);
  const [date, setDate] = useState(message.date);

  async function submit(e) {
    e.preventDefault();
    console.log("Submitting", { recipient, subject, content, date });
    await fetch(`/api/messages/${message.id}`, {
      method: "PUT",
      body: JSON.stringify({ recipient, subject, content, date }),
      headers: {
        "Content-Type": "application/json",
      },
    });
  }

  return (
    <form onSubmit={submit}>
      <h1>Edit an existing message ({subject})</h1>
        <InputField
            label={"Recipient"}
            value={recipient}
            onChangeValue={setRecipient}
        />
      <InputField
        label={"Subject"}
        value={subject}
        onChangeValue={setSubject}
      />
      <InputField
        label={"Content"}
        value={content}
        onChangeValue={setContent}
      />
      <InputField
        label={"Date"}
        value={date}
        onChangeValue={setDate}
        type="date"
      />
      <button>Submit</button>
    </form>
  );
}

export function EditMessage({ messageApi }) {
  const { id } = useParams();

  const { data: message, loading, error, reload } = useLoading(
    async () => await messageApi.getMessage(id),
    [id]
  );

  if (error) {
    return <ErrorView error={error} reload={reload()} />;
  }

  if (loading || !message) {
    return <LoadingView />;
  }

  return <EditMessageForm message={message} />;
}
