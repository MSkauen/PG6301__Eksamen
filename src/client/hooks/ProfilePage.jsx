import React from "react";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";
import { fetchJson } from "../lib/http";

export function ProfilePage() {
  const { data, error, loading, reload } = useLoading(() =>
    fetchJson("/api/profile", {
      method: "POST",
      body: JSON.stringify({}),
      headers: {
        "Content-Type": "application/json",
      },
    })
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }
  if (loading || !data) {
    return <LoadingView />;
  }

  const { username } = data;

  return (
    <div>
      <h1>Your profile</h1>
      <div>Username: {username}</div>
    </div>
  );
}
