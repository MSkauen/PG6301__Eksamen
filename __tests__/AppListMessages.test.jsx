import TestRenderer from "react-test-renderer";

import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { AppListMessages } from "../src/client/hooks/AppListMessages";

const messageApi = {
  listMessages: async () => [{ id: 1, subject: "nyMelding" }],
};

describe("message list page", () => {
  it("show messages", async () => {
    let view;
    await TestRenderer.act(async () => {
      view = TestRenderer.create(<AppListMessages messageApi={messageApi} />);
    });
    expect(view.toJSON()).toMatchSnapshot();
    expect(view.root.findByType("li").children[0]).isContaining("nyMelding");
  });

  it("show messages on dom", async () => {
    const container = document.createElement("div");
    document.body.appendChild(container);
    await act(async () => {
      ReactDOM.render(<AppListMessages messageApi={messageApi} />, container);
    });

    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("li").textContent).toEqual("nyMelding");
  });
});
