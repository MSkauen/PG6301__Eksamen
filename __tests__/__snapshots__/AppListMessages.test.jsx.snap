// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message list page show message content on dom 1`] = `"<h1>List messages</h1><li><a href=\\"/messages/1/edit\\"><b>Ola</b> [Subject:Hello world]</a></li>"`;

exports[`message list page show message sender on dom 1`] = `"<h1>List messages</h1><li><a href=\\"/messages/1/edit\\"><b>Ola</b> [Subject:Hello world]</a></li>"`;
